### PR TITLE
use tidb_kv_read_timeout as first kv request timeout (#919)

### PR DESCRIPTION
**Pull request by crazycs520**
cherry-pick #919

### CODE REVIEW

File: internal/locate/region_request.go [hunk 1]
```diff
@@ -436,7 +446,7 @@ func (state *tryFollower) next(bo *retry.Backoffer, selector *replicaSelector) (
 }
 
 func (state *tryFollower) onSendSuccess(selector *replicaSelector) {
-	if !state.fallbackFromLeader {
+	if state.fromOnNotLeader {
```
[the remark this answers is not in the document]
Comment (you06) [reply internal/locate/region_request.go:449]: The flag is different between 6.5 and master, I think here is to unify them.

File: internal/locate/region_request.go [hunk 2]
```diff
@@ -377,17 +379,21 @@ type tryFollower struct {
 	leaderIdx          AccessIndex
 	lastIdx            AccessIndex
 	labels             []*metapb.StoreLabel
+	// fromOnNotLeader indicates whether the state is changed from onNotLeader.
+	fromOnNotLeader bool
```
Comment (you06) [inline internal/locate/region_request.go:383]: By adding fromOnNotLeader, it's better to remove `state.fallbackFromLeader` flag

File: internal/locate/region_request.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (cfzjywxk) [reply internal/locate/region_request.go:449]: But L429 they are still used seperately, it's confusing.

File: internal/locate/region_request.go (hunk 2 above)
Comment (you06) [reply internal/locate/region_request.go:383]: And the flag should also set in these places.

https://github.com/tikv/client-go/blob/6f2442951b2bf717081ff383e3164f84675bf3d9/internal/locate/region_request.go#L340

https://github.com/tikv/client-go/blob/6f2442951b2bf717081ff383e3164f84675bf3d9/internal/locate/region_request.go#L355

File: internal/locate/region_request3_test.go [hunk 3]
```diff
@@ -1190,3 +1194,84 @@ func (s *testRegionRequestToThreeStoresSuite) TestStaleReadFallback2Follower() {
...
+			s.Equal(1, len(s.regionRequestSender.Stats))
+			if staleRead {
+				rpcNum := s.regionRequestSender.Stats[tikvrpc.CmdGet].Count
+				s.True(rpcNum == 1 || rpcNum == 2) // 1 rpc or 2 rpc
+			} else {
+				s.Equal(int64(3), s.regionRequestSender.Stats[tikvrpc.CmdGet].Count) // 3 rpc
+				s.Equal(3, len(reqTargetAddrs))                                      // each rpc to a different store.
+			}
+			s.Equal(0, bo.GetTotalBackoffTimes()) // no backoff since fast retry.
+			// warn: must rest MaxExecutionDurationMs before retry.
```
Comment (you06) [inline internal/locate/region_request3_test.go:1257]:
```suggestion
			// warn: must reset MaxExecutionDurationMs before retry.
```

I guess it's "reset".

File: internal/locate/region_request.go [hunk 4]
```diff
@@ -436,7 +446,7 @@ func (state *tryFollower) next(bo *retry.Backoffer, selector *replicaSelector) (
 }
 
 func (state *tryFollower) onSendSuccess(selector *replicaSelector) {
-	if !state.fallbackFromLeader {
+	if !state.fallbackFromLeader || state.fromOnNotLeader {
```
Comment (cfzjywxk) [inline internal/locate/region_request.go:449]: Should it be `!state.fallbackFromLeader && state.fromOnNotLeader`?

**Comment by cfzjywxk** [comment]
@you06 PTAL

File: internal/locate/region_request.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (crazycs520) [reply internal/locate/region_request.go]: Now, this logic is keep consistent with the master branch after https://github.com/tikv/client-go/pull/952.

File: internal/locate/region_request.go [hunk 5]
```diff
@@ -426,17 +434,17 @@ func (state *tryFollower) next(bo *retry.Backoffer, selector *replicaSelector) (
...
 	}
-	if state.fallbackFromLeader {
-		staleRead := false
-		rpcCtx.contextPatcher.staleRead = &staleRead
+	if !state.fromAccessKnownLeader {
 		replicaRead := true
 		rpcCtx.contextPatcher.replicaRead = &replicaRead
 	}
+	staleRead := false
+	rpcCtx.contextPatcher.staleRead = &staleRead
```
Comment (cfzjywxk) [inline internal/locate/region_request.go:442]:
```suggestion
	if !state.fromAccessKnownLeader {
		staleRead := false
	        rpcCtx.contextPatcher.staleRead = &staleRead
	}
	replicaRead := true
	rpcCtx.contextPatcher.replicaRead = &replicaRead
```

Should it be like this?

**Comment by cfzjywxk** [comment]
@crazycs520 
It seems the DCO fails, commits need to be signed with "-s" in git.

File: internal/locate/region_request.go (hunk 5 above)
Comment (you06) [reply internal/locate/region_request.go:442]: No, if it's from `accessKnownLeader`, replica read should not be set.

Comment (crazycs520) [reply internal/locate/region_request.go:442]: The logic is consistent with master branch. https://github.com/tikv/client-go/blob/master/internal/locate/region_request.go#L461

**Review by bufferflies — APPROVED** [review]

**Pull request merged**